src/controller_examples/rabbitmq_controller/proof/liveness/proof.rs

```diff
@@ -0,0 +1,319 @@
+    leads_to_trans_temp(assumption_and_invariants_of_all_phases(rabbitmq), true_pred(), always(lift_state(sub_resource_state_matches(SubResource::ServerConfigMap, rabbitmq))), always(lift_state(sub_resource_state_matches(SubResource::StatefulSet, rabbitmq))))
+}
+
+proof fn lemma_true_leads_to_state_matches_for_all_resources(rabbitmq: RabbitmqClusterView)
+    requires
+        rabbitmq.well_formed(),
+    ensures
+        forall |sub_resource: SubResource|
+            assumption_and_invariants_of_all_phases(rabbitmq)
+            .entails(
+                true_pred().leads_to(lift_state(#[trigger] sub_resource_state_matches(sub_resource, rabbitmq)))
+            ),
+{
+    let spec = assumption_and_invariants_of_all_phases(rabbitmq);
+
+    assert forall |action: ActionKind, sub_resource: SubResource| #![auto] spec.entails(always(lift_state(RMQCluster::pending_req_in_flight_or_resp_in_flight_at_reconcile_state(rabbitmq.object_ref(), at_step_closure(RabbitmqReconcileStep::AfterKRequestStep(action, sub_resource)))))) by {
+        use_tla_forall(spec, |step: (ActionKind, SubResource)| always(lift_state(RMQCluster::pending_req_in_flight_or_resp_in_flight_at_reconcile_state(rabbitmq.object_ref(), at_step_closure(RabbitmqReconcileStep::AfterKRequestStep(step.0, step.1))))), (action, sub_resource));
+    }
+
+    // The use of termination property ensures spec |= true ~> reconcile_idle.
+    terminate::reconcile_eventually_terminates(spec, rabbitmq);
+    // Then we can continue to show that spec |= reconcile_idle ~> []current_state_matches(rabbitmq).
+
+    // The following two lemmas show that spec |= reconcile_idle ~> init /\ no_pending_req.
+    lemma_from_reconcile_idle_to_scheduled(spec, rabbitmq);
+    lemma_from_scheduled_to_init_step(spec, rabbitmq);
+
+    // After applying this lemma, we get spec |= init /\ no_pending_req ~> create_headless_service /\ pending_req.
+    lemma_from_init_step_to_after_create_headless_service_step(spec, rabbitmq);
+
+    use_tla_forall(spec, |res: SubResource| always(lift_state(helper_invariants::every_resource_create_request_implies_at_after_create_resource_step(res, rabbitmq))), SubResource::ServerConfigMap);
+    assert forall |sub_resource: SubResource| sub_resource != SubResource::StatefulSet implies spec.entails(
+        lift_state(#[trigger] pending_req_in_flight_at_after_get_resource_step(sub_resource, rabbitmq))
+            .leads_to(lift_state(pending_req_in_flight_at_after_get_resource_step(next_resource_get_step_and_request(rabbitmq, sub_resource).0.get_AfterKRequestStep_1(), rabbitmq)))) by {
+        use_tla_forall_for_sub_resource(spec, sub_resource, rabbitmq);
+        let next_resource = next_resource_get_step_and_request(rabbitmq, sub_resource).0.get_AfterKRequestStep_1();
+        lemma_from_after_get_resource_step_to_resource_matches(spec, rabbitmq, sub_resource, next_resource);
+    }
+    // Thanks to the recursive construction of macro.
```

Comment:
coz it gives your not only `a ~> d`, but also `a ~> b`, `a ~> c`?

Reply:
Yes. But of course we can use `leads_to_trans_temp` n times...

Reply:
But if we (are able to) write a function for leads_to_trans_n, usually the post condition will only contain `p1 ~> pn`.

Reply:
yes